### ISSUE TITLE
PR builds checkout the wrong commit if branch is updated between queuing and running the build

### ISSUE THREAD

**Issue by peterjwest** [travis-ci/travis-ci#7459]
I'm having occasional problems with PR builds, where the reported commit is not the commit checked out in the build.

After seeing a build which seemed to use the wrong commit, I reran the build in debug mode:
![screen shot 2017-03-10 at 18 29 27](https://cloud.githubusercontent.com/assets/246023/23870333/02fb2474-081e-11e7-9298-ae9485c62d9b.png)

SSHing into the build, this is what git told me:
![screen shot 2017-03-10 at 18 30 18](https://cloud.githubusercontent.com/assets/246023/23870357/160eb0b2-081e-11e7-9b25-3434224f34c8.png)

This has happened to me twice over 3 days / approx 20 builds.

I amended the timestamp of the commit and force pushed, which fixed the issue.

Link to build:
https://travis-ci.com/diffblue/application-server/builds/41645951

The cause may be the `git fetch origin +refs/pull/[PR]/merge:` command, since this will always get the latest commit, regardless of the specific commit that is meant to be tested.

**Comment by peterjwest** [comment]
My workaround is this:
```
before_install:
  # Check commit matches expected commit (because of Travis bug)
  - |
    if [[ "$TRAVIS_COMMIT" != "$(git rev-parse HEAD)" ]]; then
      echo "Commit $(git rev-parse HEAD) doesn't match expected commit $TRAVIS_COMMIT"
    fi
```

This simply fails the build if the commit does not match the expected one, which isn't great, but its better than us merging failing code into master.

**Comment by BanzaiMan** [comment]
I am not sure if we can overcome this sort of race condition.

We configure a build with the payload handed to us by GitHub. If the merge commit is no longer valid when the build runs, there is nothing we can do.

This is basically the same thing when you restart a really old build. Since our builds are typically configured with the default depth of 50, if there have been more than 50 commits since the build initially ran, the restart will fail.

I'm inclined to close this.

**Comment by peterjwest** [comment]
I'm astounded by this attitude, this is a commercial platform that people test and build critical software on, and you're happy with "this is probably the right commit, but if it's not we won't even warn you".

Here's a possible solution: when the build is triggered, store two commit hashes (one for the base branch, one for the merging branch) merge these as part of the build. If the merge fails, the build would be blocked on Github anyway.

**Comment by BanzaiMan** [comment]
We will be announce a new feature that eases this sort of pain: When there is a new push to a branch, we will optionally cancel previous builds if they have not started yet. Please see https://github.com/travis-ci/beta-features/issues/8.

Thanks.

**Comment by BanzaiMan** [comment]
Can this be closed?

**Comment by peterjwest** [comment]
Cancelling previous builds mitigates this problem, but I wouldn't say I'm happy with this as a solution.

When something fails in CI it's extremely useful to find out when that change was introduced. Having a history previous builds is an invaluable way of establishing that. With the current set up I can't even re-run previous builds because they will always build the latest version.

**Comment by hisabimbola** [comment]
I am currently facing something similar to this.

If I restart old builds, it fails and that's because it clones repository with a limited depth. Might be better if we can clone a commit history instead.

**Comment by rpkilby** [comment]
btw, this looks to be a duplicate of #6337.

**Comment by peterjwest** [comment]
Thanks, closing.

**Comment by OliverJAsh** [comment]
Is it not possible for Travis to checkout the specific commit that's described at the top of the build page? For example:

![image](https://user-images.githubusercontent.com/921609/76958825-4ff3ea80-6910-11ea-9be3-32da06b843a6.png)

Then, instead of running…

```bash
git fetch origin +refs/pull/4812/merge:
git checkout -qf FETCH_HEAD
```

… Travis would run:

```bash
git fetch origin a50c7a8ec7ccbbd4655c908028f32debeaa87b69
git checkout -qf a50c7a8ec7ccbbd4655c908028f32debeaa87b69
```

/cc @BanzaiMan @peterjwest 

---

> When there is a new push to a branch, we will optionally cancel previous builds if they have not started yet.

This only helps when the previous build hasn't started yet. In my experience, the previous build has almost always started by the time we push a new commit, in which case cancellation can't help us.

**Comment by OliverJAsh** [comment]
This issue also happens even when the branch is not updated. It seems that Travis (or GitHub?) generates two merge commits, which are identical, but one is generated a second later. Thus, we run into the same issue where the checked out commit is different from the commit specified at the top of the build page.

Potentially related issues:

- https://github.com/travis-ci/travis-ci/issues/10210
- https://github.com/travis-ci/travis-ci/issues/10224

**Comment by peterjwest** [comment]
Sorry I don't really remember, we definitely still had issues with it.
